src/drone.rs

```diff
@@ -0,0 +1,340 @@
+        // [WIP] This is placeholder code for a proper rate limiter.
+        // Right now it will only allow one total drone request per IP
+        if self.ip_cache.contains(&ip) {
+            // Add proper error handling here
+            Err(ip)
+        } else {
+            self.add_ip_to_cache(ip);
+            Ok(ip)
+        }
+    }
+
+    pub fn send_airdrop(&mut self, req: DroneRequest) -> Result<usize, io::Error> {
+        self.request_current += 1;
+
+        if self.check_request_count() {
+            let tx: Transaction;
+            let requests_socket = UdpSocket::bind("0.0.0.0:0").unwrap();
+            let transactions_socket = UdpSocket::bind("0.0.0.0:0").unwrap();
+
+            let mut client = ThinClient::new(
+                self.requests_addr,
+                requests_socket,
+                self.transactions_addr,
+                transactions_socket,
+            );
+            let last_id = client.get_last_id();
+
+            match req {
+                DroneRequest::GetAirdrop {
+                    request_type,
+                    client_public_key,
+                } => match request_type {
+                    DroneRequestType::SmallBatch => {
+                        tx = Transaction::new(
+                            &self.mint_keypair,
+                            client_public_key,
+                            SMALL_BATCH,
+                            last_id,
+                        );
+                    }
+                    DroneRequestType::TPSBatch => {
+                        tx = Transaction::new(
+                            &self.mint_keypair,
+                            client_public_key,
+                            TPS_BATCH,
+                            last_id,
+                        );
+                    }
+                },
+            }
+            client.transfer_signed(tx)
+        } else {
+            Err(Error::new(ErrorKind::Other, "request limit reached"))
+        }
+    }
+}
+
+#[cfg(test)]
+mod tests {
+    use bank::Bank;
+    use crdt::{get_ip_addr, TestNode};
+    use drone::{Drone, DroneRequest, DroneRequestType, REQUEST_CAP, SMALL_BATCH, TIME_SLICE,
+                TPS_BATCH};
+    use logger;
+    use mint::Mint;
+    use server::Server;
+    use signature::{KeyPair, KeyPairUtil};
+    use std::io::sink;
+    use std::net::{SocketAddr, UdpSocket};
+    use std::sync::Arc;
+    use std::sync::atomic::{AtomicBool, Ordering};
+    use std::thread::sleep;
+    use std::time::Duration;
+    use thin_client::ThinClient;
+
+    #[test]
+    fn test_check_request_count() {
+        let keypair = KeyPair::new();
+        let mut addr: SocketAddr = "0.0.0.0:9900".parse().unwrap();
+        addr.set_ip(get_ip_addr().unwrap());
+        let mut transactions_addr = addr.clone();
+        transactions_addr.set_port(8000);
+        let mut requests_addr = addr.clone();
+        requests_addr.set_port(8003);
+        let mut drone = Drone::new(
```

Comment:
Can you add a `new_from_server_addr` helper that assumes `transactions_addr` with be port + 0 and `requests_addr` will be port + 3?  Too many 8000/8003 magic numbers in these tests.